packages/adapter-node/src/index.js

```diff
@@ -6,10 +6,56 @@ export const path = env('SOCKET_PATH', false);
 export const host = env('HOST', '0.0.0.0');
 export const port = env('PORT', !path && '3000');
 
+const listen_pid = parseInt(env('LISTEN_PID', '-1'));
+const listen_fds = parseInt(env('LISTEN_FDS', '-1'));
+const timeout = parseInt(env('TIMEOUT', '-1'));
+const systemd_socket_fd = 3;
+
 const server = polka().use(handler);
 
-server.listen({ path, host, port }, () => {
-	console.log(`Listening on ${path ? path : host + ':' + port}`);
-});
+function close() {
+	// @ts-expect-error this was added in 18.2.0 but is not reflected in the types
+	server.server.closeIdleConnections();
+	server.server.close();
+}
+
+if (listen_pid === process.pid && listen_fds === 1) {
+	server.listen({ fd: systemd_socket_fd }, () => {
+		console.log('Listening on file descriptor 3');
+	});
+
+	if (timeout) {
+		/** @type {NodeJS.Timeout | void} */
+		let timeout_id;
+		let requests = 0;
+
+		/** @param {import('node:http').IncomingMessage} req */
+		function on_request(req) {
+			requests++;
+
+			if (timeout_id) {
+				timeout_id = clearTimeout(timeout_id);
+			}
+
+			req.on('close', on_request_close);
+		}
+
+		function on_request_close() {
+			requests--;
+
+			if (requests === 0) {
+				timeout_id = setTimeout(close, timeout * 1000);
+			}
+		}
+
+		server.server.on('request', on_request);
+	}
+} else {
+	server.listen({ path, host, port }, () => {
+		console.log(`Listening on ${path ? path : host + ':' + port}`);
+	});
+}
+
+process.on('SIGTERM', close);
```

Comment:
This SIGTERM handler is getting added in both cases, even if the user is not using this new feature. This may be desirable, but I think it deserves an explanation.

Reply:
We had a discussion about graceful shutdown on Discord and in this issue https://github.com/sveltejs/kit/issues/9540 I thought I might as well just include this line. Should I remove it so we can discuss this a bit more first?

Reply:
I think it's a reasonable change. I don't have an especially strong opinion about the atomicity of PRs for this, but we should mark this PR as resolving that issue, and we should mention it in the changelog.

Reply:
The issue also mentions SIGINT, not just SIGTERM, I'm realizing. Were there any discussions about that?

Reply:
I've only included `SIGTERM` because I know this is what systemd sends when it wants to stop a service. But I've just verified that `SIGNINT` is the signal you receive when hitting `Ctrl + C` in the terminal to close the app. So I should add a handler for this as well. The docs currently have [a section about handling these signals yourself](https://kit.svelte.dev/docs/adapter-node#troubleshooting-is-there-a-hook-for-cleaning-up-before-the-server-exits). I wonder if we need to change anything about that.

Reply:
Hm. If we're already mentioning that in the docs, maybe we _shouldn't_ add those handlers here? Could multiple handlers interfere with each other if someone were already attaching their own elsewhere? Could adding the handlers break anyone's apps who were depending on whatever the default behavior was?

Reply:
I don't think it would be an issue because all this does is tell the server to handle all remaining request and then gracefully shutdown. If the user has other code to run on shutdown multiple handlers still work. There is this scenario where you might want to run some code after the server was closed. For this you would need to pass a callback to `server.server.close(cb)` and that's not possible because the adapter doesn't expose it. But since the server instance was never exposed by the adapter in the first place I don't think there is a conflict.

I think in general you would always want the server to close all idle connections and handle all remaining requests before shutting down. If we don't include this then the only way to do this is make use of the custom server feature. But I would argue that handling this is probably the better default in the majority of cases.

Reply:
If we include this we probably want to add one or two sentences and mention that the adapter calls `closeIdleConnections` and `close` on `SIGTERM` and `SIGINT`.

Reply:
This all makes sense to me, but I am coming from a place of ignorance so I'll refrain from approving the PR and leave that to @Conduitry

Reply:
Yeah, that all sounds reasonable as a default. If users _really_ need something else, they always have the custom entrypoint escape hatch.

We should mention this in the changelog and update the docs, though.

Reply:
I wrote a bit about how the graceful shutdown fill work here https://github.com/sveltejs/kit/issues/9540#issuecomment-1902072756